### PR TITLE
Automatically add node labels beta.kubernetes.io/{os,arch}

### PR DESCRIPTION
**Pull request by luxas**
Proposal: #17981
As discussed in #22623:

> @davidopp: #9044 says cloud provider but can also cover platform stuff.

Adds a label `beta.kubernetes.io/platform` to `kubelet` that informs about the os/arch it's running on.
Makes it easy to specify `nodeSelectors` for different arches in multi-arch clusters.

``` console
$ kubectl get no --show-labels
NAME        STATUS    AGE       LABELS
127.0.0.1   Ready     1m        beta.kubernetes.io/platform=linux-amd64,kubernetes.io/hostname=127.0.0.1
$ kubectl describe no
Name:           127.0.0.1
Labels:         beta.kubernetes.io/platform=linux-amd64,kubernetes.io/hostname=127.0.0.1
CreationTimestamp:  Thu, 31 Mar 2016 20:39:15 +0300
```

@davidopp @vishh @fgrzadkowski @thockin @wojtek-t @ixdy @bgrant0607 @dchen1107 @preillyme

### CODE REVIEW

**Comment by davidopp** [comment]
Would it be better to split it into two labels, one for OS and one for arch?

**Comment by luxas** [comment]
Do you have a specific case where it would be useful?
To me, it doesn't matter so much, but I think we're fine with one, cause in most cases, you must specify both os and arch

**Comment by luxas** [comment]
@davidopp Is it ok as-is or should I change something?
I would like to hear what you think.

I lean towards just one label as implemented here.
It would make less noise, and it's only one label to specify in `nodeSelector`
And I honestly can't come up with a case where you only would specify os or arch.

WDYT about `os-arch` vs `os_arch` then? 
`os/arch` (which I would have preferred) wasn't allowed as a value.

**Comment by vishh** [comment]
I have a preference for separating OS from Arch as well. Depending on the
use case this PR is trying to address, I guess the value of the split might
not be obvious.

**Comment by ixdy** [comment]
Note prior art: with #19905, the server version string includes a field "platform" which is `fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH)`.

**Comment by luxas** [comment]
@davidopp @vishh Updated to two labels for reasons no one knows today but may prove useful in the future.
PTAL

**Comment by davidopp** [comment]
LGTM

**Comment by vishh** [comment]
@davidopp @dchen1107 @bgrant0607 What is the purpose of `NodeStatus.NodeSystemInfo`? Is that supposed to be a super set of all the node related labels that are exposed by the kubelet? Will users be able to use the fields in NodeSystemInfo for scheduling in the future?

**Comment by vishh** [comment]
@davidopp can we hold on to this PR until https://github.com/kubernetes/kubernetes/pull/23684#issuecomment-210153386 is resolved?

**Comment by davidopp** [comment]
Removed LGTM.

Yeah, I don't know what the intention of NodeSystemInfo was, but I would argue that the fields that people would conceivably want to write scheduling constraints against (KernelVersion, OSImage, ContainerRuntimeVersion) should be labels.

**Comment by dchen1107** [comment]
@vishh I don't think NodeStatus.NodeSystemInfo is the superset of all the node related label exposed by Kubelet. Instead, my opinion is that all fields in NodeSystemInfo should be automatically exposed as labels.

**Comment by davidopp** [comment]
BTW test failure is
```
Verifying ./hack/../hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./pkg/kubelet/kubelet.go
FAILED   ./hack/../hack/verify-gofmt.sh	6s
```

**Comment by luxas** [comment]
So, what should I do? 
Should I add a lot of other labels in this PR for scheduling or will this happen automagically with an other approach?
Updating generated docs

**Comment by davidopp** [comment]
@vishh and I discussed today and we agree with @dchen1107 , the best thing to do is to just publish everything in NodeStatus.NodeSystemInfo as a label, so we don't have to do work in the future to keep the label publishing in sync with the NodeStatus.NodeSystemInfo. I'm not an expert on reflection but hopefully you can do this transparently, i.e. use the json field tag as the key and (obviously) the value as the value. The one thing we weren't sure is whether this logic to reflect NodeStatus.NodeSystemInfo into labels should be done in Kubelet or NodeController.

**Comment by luxas** [comment]
I'm OK with that, but I don't see that os/arch is exposed there:
```go
// NodeSystemInfo is a set of ids/uuids to uniquely identify the node.
type NodeSystemInfo struct {
	// Machine ID reported by the node.
	MachineID string `json:"machineID"`
	// System UUID reported by the node.
	SystemUUID string `json:"systemUUID"`
	// Boot ID reported by the node.
	BootID string `json:"bootID"`
	// Kernel Version reported by the node from 'uname -r' (e.g. 3.16.0-0.bpo.4-amd64).
	KernelVersion string `json:"kernelVersion"`
	// OS Image reported by the node from /etc/os-release (e.g. Debian GNU/Linux 7 (wheezy)).
	OSImage string `json:"osImage"`
	// ContainerRuntime Version reported by the node through runtime remote API (e.g. docker://1.5.0).
	ContainerRuntimeVersion string `json:"containerRuntimeVersion"`
	// Kubelet Version reported by the node.
	KubeletVersion string `json:"kubeletVersion"`
	// KubeProxy Version reported by the node.
	KubeProxyVersion string `json:"kubeProxyVersion"`
}
```
Haven't used reflection too much, but I may see if I can do something.
Should we merge this one first and update the api later or how do we proceed?

**Comment by davidopp** [comment]
Hmm, good point. arch is definitely not in NodeSystemInfo. OSImage is vaguely like GOOS but not exactly the same.

@dchen1107 @vishh thoughts?

**Comment by luxas** [comment]
`OSImage` doesn't make sense in dockerized setups, it's always `Debian` there (even on other platforms)

Running on my `amd64` machine:
```console
$ uname -a
Linux THENINJA 4.2.0-25-generic #30-Ubuntu SMP Mon Jan 18 12:31:50 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

## Qemu emulation
$ docker run -it gcr.io/google_containers/hyperkube-arm:v1.3.0-alpha.2 /bin/bash
root@970a2fc30c74:/# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
root@970a2fc30c74:/# uname -a
Linux 970a2fc30c74 4.2.0-25-generic #30-Ubuntu SMP Mon Jan 18 12:31:50 UTC 2016 armv7l GNU/Linux
```

**Comment by luxas** [comment]
Side note: @ixdy said above that #19905 added `platform` to `pkg/version` `Info.platform`, but that's just used for `kubectl` AFAIK

**Comment by vishh** [comment]
`OSImage` is already there. Let's add `Architecture` to `NodeSystemInfo`

> OSImage doesn't make sense in dockerized setups, it's always Debian there (even on other platforms)

This is one more reason why kubelet does not fit the docker container model. `rkt fly` is a better way to run kubelet IMHO.

**Comment by davidopp** [comment]
@vishh I don't understand your previous comment ("OSImage is already there"). Why would we want an OS label that always says Debian? I assume the user wants whatever GOOS returns (as in the original version of this PR).

**Comment by thockin** [comment]
I'm not sure how I feel about this.  Won't this bloat the objects
(dupplicate info) and obscure user-meaningful labels?  Is our intention to
make labels the ONLY selection?  We have spacial-cased other selections
(e.g. NodeName) - should that become a label too?

**Comment by vishh** [comment]
@davidopp 
Apologies. My comment was not relevant to this PR. It was a more generic PR. What I meant was we should add the missing fields to NodeStatusInfo.

@thockin: @davidopp mentioned that labels will be the only selecting mechanism for the scheduler.

**Comment by davidopp** [comment]
@thockin I think it's simpler for users if we say scheduling is based on resources and labels and nothing else. (Taints and tolerations will also come into play but I would like those to be system-generated and not something users have to think about.) If we say you can also constrain on random fields of status objects I think it will get confusing. So I think the reasonable choices are to automatically reflect everything from NodeSystemInfo as labels in the "system" (kubernetes.io) namespace, or manually publish the labels of specific interest (like the original version of this PR).

@vishh So IIUC you are advising that we add GOOS and GOARCH to NodeSystemInfo (using friendlier field names than that, of course), and then automatically publish everything from NodeSystemInfo as labels?

**Comment by thockin** [comment]
Why have them as status at all, then?

**Comment by davidopp** [comment]
I assume it's so it shows up in "kubectl describe node" (at least that seems to be the only for that info today). So yeah, if we put it in labels, maybe we don't need it in status?

**Comment by vishh** [comment]
AFAIK the contents in node status were meant to help with automating machine management in the future.

@davidopp: yes.

**Comment by luxas** [comment]
Should this one be modified to add ´Info.Architecture` and `Info.OperatingSystem` first? 
Then, in a follow-up, we may continue with exposing `Info.*` as labels.

**Comment by davidopp** [comment]
Yes, I think it's reasonable to change this PR so that it just exposes GOOS and GOARCH as new fields of Info, and then do the automatic publication of all Info fields as labels in a separate PR.

**Comment by dchen1107** [comment]
I am ok to add arch as new field of NodeSystemInfo in this pr, and expose them in a separate PR.

I agreed with @thockin on the concern of the duplicate data. But solving this by moving a bunch of fields in NodeSystemInfo to labels is a big no since the labels is not well structured data, and don't show up in the API documentation. I also want Kubelet / Node is the source of the truth of those NodeSystemInfo, on top of which, the user can build other control systems.

**Comment by bgrant0607** [comment]
I would prefer that we not go wild exposing all these fields as labels.

**Comment by bgrant0607** [comment]
Is there any reason to expose more than architecture right now?

And even architecture: do people want to run mixed-architecture clusters?

**Comment by davidopp** [comment]
> And even architecture: do people want to run mixed-architecture clusters?

Original motivation for OS and architecture is here:
https://github.com/kubernetes/kubernetes/issues/22623#issuecomment-203574605
(Also mentioned in #17981)

I think it's fine to say we won't blanket expose everything in NodeSystemInfo (though I think it has the benefit of being possible to do automatically using reflection) but I think there's a strong case to be made for OS and arch.

File: pkg/api/unversioned/well_known_labels.go [hunk 1]
```diff
@@ -20,3 +20,5 @@ const LabelHostname = "kubernetes.io/hostname"
 const LabelZoneFailureDomain = "failure-domain.beta.kubernetes.io/zone"
 const LabelZoneRegion = "failure-domain.beta.kubernetes.io/region"
 const LabelInstanceType = "beta.kubernetes.io/instance-type"
+const LabelOs = "beta.kubernetes.io/os"
```
Comment (bgrant0607) [inline pkg/api/unversioned/well_known_labels.go]: LabelOS

**Comment by bgrant0607** [comment]
where OS is just linux (as opposed to Debian, etc.) or windows?

**Comment by luxas** [comment]
Okay, so what's the conclusion here?
We should expose just selected ones from `NodeSystemInfo`.
We should expose `os/arch`
Should `OperatingSystem` and `Architecture` be added to `NodeSystemInfo`?

And what should be done in this PR?

**Comment by davidopp** [comment]
https://github.com/kubernetes/kubernetes/issues/6079#issuecomment-212735350
is a use case for exposing something like Kubelet version as a label

**Comment by vishh** [comment]
@luxas 

> Should OperatingSystem and Architecture be added to NodeSystemInfo?

There is no confusion about this part. So let's make that change first.

**Comment by luxas** [comment]
Added `Architecture` and `OperatingSystem` to `NodeSystemInfo` for `api` and `api/v1`.
`kubectl describe no` shows these fields under status.
And the labels are created at node creation.

**Comment by luxas** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by luxas** [comment]
There's something wrong with `hack/update-codecgen.sh` and `hack/verify-codecgen.sh`
When verifying:
```
./cmd/libs/go2idl/client-gen/testdata/apis/testgroup.k8s.io/types.generated.go is out of date. Please run hack/update-codecgen.sh
```
but when I'm updating, nothing changes. What may cause that?

BTW, I'm on a one-week journey from tomorrow, so this will have to wait for that. 
Feel free to give feedback anyway.

**Comment by luxas** [comment]
Splitted into two commits for easier reviewing.

**Comment by davidopp** [comment]
This PR looks fine to me.

@luxas can you rebase and run `hack/update-all.sh?` The test is failing with

```
./cmd/libs/go2idl/client-gen/testdata/apis/testgroup.k8s.io/types.generated.go is out of date. Please run hack/update-codecgen.sh
Removing /tmp/codecgen.M94uo0
FAILED   ./hack/../hack/verify-codecgen.sh	45s
```

**Comment by luxas** [comment]
Even though I rebase, the `codecgen` problem persists.
`hack/update-codecgen.sh` produces files `hack/verify-codecgen.sh` doesn't like.

It would be good to get this in before `v1.3.0-alpha.3` @david-mcmahon

**Comment by david-mcmahon** [comment]
@ihmccreery, an interest has been expressed here to get this into the next alpha.  Sorry for the busy week!

**Comment by davidopp** [comment]
It hasn't been LGTMd yet.

I'll try to get to it in the next 24 hours.

**Comment by davidopp** [comment]
Oh sorry, I forgot the history on this one. The PR is fine, the question is why hack/verify-codecgen.sh is failing. Can you rebase, run hack/update-all.sh again, and if the tests still fail, we can ask @lavalamp or @wojtek-t what's going on...

**Comment by ihmccreery** [comment]
`alpha.3` was already tagged.  I'm encountering local complications building (my workstation's out of space, still trying to figure out why), but the tag's sunk; target `alpha.4`.

**Comment by davidopp** [comment]
ping @luxas : Can you rebase and run hack/update-all.sh again? If that doesn't fix it, we can investigate more.

**Comment by luxas** [comment]
@davidopp Soon I'll rebase and `hack/update-all.sh` a third round.
The two earlier times, it updated successfully and everything was fine except from that `verify-codecgen` failed.

**Comment by luxas** [comment]
@davidopp @lavalamp @wojtek-t Why is `hack/update-all.sh` updating `codecgen` incorrectly for me:
```console
$ hack/update-all.sh
$ hack/verify-codecgen.sh
-		yyj9++
-		if yyhl9 {
-			yyb9 = yyj9 > l
+		yyj8++
+		if yyhl8 {
+			yyb8 = yyj8 > l
 		} else {
-			yyb9 = r.CheckBreak()
+			yyb8 = r.CheckBreak()
 		}
-		if yyb9 {
+		if yyb8 {
 			break
 		}
 		z.DecSendContainerState(codecSelfer_containerArrayElem1234)
-		z.DecStructFieldNotFound(yyj9-1, "")
+		z.DecStructFieldNotFound(yyj8-1, "")
 	}
 	z.DecSendContainerState(codecSelfer_containerArrayEnd1234)
 }
./cmd/libs/go2idl/client-gen/testdata/apis/testgroup.k8s.io/types.generated.go is out of date. Please run hack/update-codecgen.sh
Removing /tmp/codecgen.JER26A
```

Note that `update` has changed the number to 9 for all variables, but `verify` wants 8.
This has happened three times in a row now and is the only blocker for this PR.

**Comment by lavalamp** [comment]
@luxas let's track the issue in #25103. I'm taking a look now.

**Comment by lavalamp** [comment]
@luxas As soon as #25108 merges, you can rebase and then you shouldn't have this problem any more. Thanks for bearing with us...

**Comment by luxas** [comment]
Argh, why does it fail now? Locally, `verify-codecgen` passes. cc @wojtek-t

**Comment by luxas** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by luxas** [comment]
ping @lavalamp 
`verify-codecgen` doesn't seem to like me :)
locally, it passes, but here it failed twice w/ the same error (run update-all)

**Comment by lavalamp** [comment]
Argh. @thockin has issues with it too. I will look shortly.

**Comment by luxas** [comment]
ping @davidopp @lavalamp 
I think it'll pass now
PTAL

**Comment by davidopp** [comment]
LGTM

**Comment by luxas** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by luxas** [comment]
@davidopp Any clue what's failing in integration?
`FAIL	k8s.io/kubernetes/pkg/kubelet	0.768s`

Removing `ok-to-merge` as it isn't needed anymore

**Comment by luxas** [comment]
Adding LGTM back now when the rebase/flake is fixed.

**Comment by luxas** [comment]
@david-mcmahon FYI, I'd really like this to get merged before `v1.3.0-alpha.4` so we can test the changes out.
But I don't know if I should add priority just for that (right now 7th in the queue)
Please keep in mind before cutting the release :)

**Comment by luxas** [comment]
@jlowdermilk FYI, this has been ín the queue now three times when the submit-queue stopped and started again. Now it got kicked out of the queue because of what I think is a flake: 
```
FAIL	k8s.io/kubernetes/contrib/mesos/pkg/offers	5.889s
```
That hasn't showed up earlier when these commits have been tested, so it have to be a flake.
Will retest unit.

**Comment by luxas** [comment]
@k8s-bot unit test this please github issue: #IGNORE

**Comment by luxas** [comment]
Adding P1 given that we want it merged before `v1.3.0-alpha.4` and because it was on the top of the queue some hours ago but was kicked out due to flakes/unstable queue

**Comment by davidopp** [comment]
@luxas if you want these also published as labels, you're running low on time to send a PR, as the code freeze for 1.3 is this Friday.

**Comment by luxas** [comment]
@davidopp I do not really understand what you mean, os/arch is already exposed as labels on every node:
```
$ hack/local-up-cluster.sh
...
$ _output/local/bin/linux/amd64/kubectl describe no
Name:			127.0.0.1
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/os=linux
			kubernetes.io/hostname=127.0.0.1
CreationTimestamp:	Mon, 16 May 2016 08:56:22 +0300
Phase:			
Conditions:
  Type		Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----		------	-----------------			------------------			------				-------
  OutOfDisk 	False 	Mon, 16 May 2016 08:57:26 +0300 	Mon, 16 May 2016 08:56:22 +0300 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  Ready 	True 	Mon, 16 May 2016 08:57:26 +0300 	Mon, 16 May 2016 08:56:23 +0300 	KubeletReady 			kubelet is posting ready status
Addresses:	127.0.0.1,127.0.0.1
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					4
 memory:				16359708Ki
 pods:					110
System Info:
 Machine ID:			dce29be2dcd94c1b958052dc77e08e6f
 System UUID:			3C4DC303-9C6C-11E4-A26A-F0761C62F136
 Boot ID:			5d120efb-67d6-4e91-9ae7-4c666fa4ab42
 Kernel Version:		4.2.0-25-generic
 OS Image:			Ubuntu 15.10
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.9.1
 Kubelet Version:		v1.3.0-alpha.1.1829+aada051b20c9e4-dirty
 Kube-Proxy Version:		v1.3.0-alpha.1.1829+aada051b20c9e4-dirty
ExternalID:			127.0.0.1
Non-terminated Pods:		(0 in total)
...
```

But thanks for notifying anyway, I'll have to be quick with some other things

**Pull request merged**